kernel/src/net/socket/ip/raw/mod.rs

```diff
@@ -0,0 +1,427 @@
+            .ok_or_else(|| Error::with_message(Errno::ENOTCONN, "the socket is not connected"))
+    }
+
+    fn sendmsg(
+        &self,
+        reader: &mut dyn MultiRead,
+        message_header: MessageHeader,
+        flags: SendRecvFlags,
+    ) -> Result<usize> {
+        // TODO: Deal with flags
+        debug_assert!(flags.is_all_supported());
+
+        let MessageHeader {
+            addr,
+            control_message,
+        } = message_header;
+
+        let remote_endpoint = match addr {
+            Some(remote_addr) => {
+                let endpoint = remote_addr.try_into()?;
+                self.try_bind_ephemeral(&endpoint)?;
+                endpoint
+            }
+            None => {
+                let ip_addr: IpEndpoint = match self.addr() {
+                    Ok(ip_addr) => ip_addr.try_into()?,
+                    Err(_) => {
+                        return Err(Error::with_message(
+                            Errno::EDESTADDRREQ,
+                            "the destination address is not specified",
+                        ))
+                    }
+                };
+                IpEndpoint::new(ip_addr.addr, 0)
+            }
+        };
+
+        if control_message.is_some() {
+            // TODO: Support sending control message
+            warn!("sending control message is not supported");
+        }
+
+        let mut op = IpHdrIncl::new();
+        let _ = self.get_option(&mut op);
+
+        // TODO: Block if the send buffer is full
+        self.try_send(reader, &remote_endpoint.addr, flags, op)
+    }
+
+    fn recvmsg(
+        &self,
+        writer: &mut dyn MultiWrite,
+        flags: SendRecvFlags,
+    ) -> Result<(usize, MessageHeader)> {
+        // TODO: Deal with flags
+        debug_assert!(flags.is_all_supported());
+
+        // This is a temporary solution. Bind to a specific iface: iface[0]
+        // TODO: Monitor multiple ifaces
+        let bind_iface_addr = IpEndpoint::new(IpAddress::v4(10, 0, 2, 15), 0);
+        self.try_bind_ephemeral(&bind_iface_addr)?;
+
+        let (received_bytes, addr) = self.recv(writer, flags)?;
+        // TODO: Receive control message
+        let IpAddress::Ipv4(ip_addr) = addr;
+        let message_header = MessageHeader::new(Some(SocketAddr::IPv4(ip_addr, 0)), None);
+        Ok((received_bytes, message_header))
+    }
+
+    fn get_option(&self, option: &mut dyn SocketOption) -> Result<()> {
+        match_sock_option_mut!(option, {
+            socket_errors: SocketError => {
+                self.options.write().socket.get_and_clear_sock_errors(socket_errors);
+                return Ok(());
+            },
+            _ => ()
+        });
+        let _ = self.options.read().socket.get_option(option);
+        self.options.read().ip.get_option(option)
+    }
+
+    fn set_option(&self, option: &dyn SocketOption) -> Result<()> {
+        if self.options.write().socket.set_option(option).is_ok() {
```

Comment:
Write `let options = self.options.write()` to avoid taking the write lock multiple times.

Same for `get_option`, where taking the read lock multiple times is also unnecessary.